### ISSUE TITLE
Flow of integer only inference

### ISSUE THREAD

**Issue by ajay-vikram** [Xilinx/brevitas#1303]
Hey, I want to understand the flow of the below quantized model for integer only inference. 

```
weight_bit_width = 4
activation_bit_width = 4

# Encoder
class QuantizedCNNModel(nn.Module):
    def __init__(self, dropout_prob=0.2):
        super(QuantizedCNNModel, self).__init__()
        self.conv1 = QuantConv2d(2, embedding_dim, kernel_size=3, stride=2, padding=1, bias=False,
                        weight_bit_width=weight_bit_width,
                        weight_quant=Int8WeightPerTensorFixedPoint,
                        bias_quant=Int32Bias,
                        return_quant_tensor=True)

        self.fc = QuantConv2d(embedding_dim, 11, kernel_size=1, bias=False,
                        weight_bit_width=weight_bit_width,
                        weight_quant=Int8WeightPerTensorFixedPoint,
                        bias_quant=Int32Bias,
                        return_quant_tensor=True)
        
        self.pool = nn.AdaptiveAvgPool2d((1, 1))
        self.quant_input = QuantIdentity(bit_width=activation_bit_width, return_quant_tensor=True)
        self.quant_output = QuantIdentity(bit_width=activation_bit_width, return_quant_tensor=False)
        self.quant_conv1 = QuantIdentity(bit_width=activation_bit_width, return_quant_tensor=True)
        
    def forward(self, x):
        x = self.quant_input(x)
        x = self.conv1(x)
        x = self.quant_conv1(x)
        logits = self.fc(x)
        logits = self.pool(logits)
        logits = logits.view(logits.size(0), -1)
        return self.quant_output(logits)
```

Once the below model is quantized and we get the .pt file for it, I have been able to extract the weights and scales but I am not able to get the same accuracy on the exact same test set. Can you help me out with the integer inference and if the scaling and operations are being done in the right way.

```
import torch
import torch.nn as nn
import torch.nn.functional as F

class InferenceModel(nn.Module):
    def __init__(self, model, checkpoint):
        super(InferenceModel, self).__init__()

        # Quantization scales
        self.inp_scale = model.quant_input.act_quant.fused_activation_quant_proxy.tensor_quant.scaling_impl.value.item()
        self.out_scale = model.quant_output.act_quant.fused_activation_quant_proxy.tensor_quant.scaling_impl.value.item()
        self.scale_conv1 =  model.conv1.quant_weight().scale
        self.scale_fc =  model.fc.quant_weight().scale
        self.act_scale_conv1 = model.quant_conv1.act_quant.fused_activation_quant_proxy.tensor_quant.scaling_impl.value.item()
        
        # Integer weights stored as float tensors (simulated quantization)
        self.weight_conv1 =  model.conv1.quant_weight().int().float()
        self.weight_fc =  model.fc.quant_weight().int().float()

        # Store stride/padding
        self.stride1 = 2
        self.pad1 = 1

    def forward(self, x):
        # Step 1: Quantize input (simulate 4-bit range [0-15])
        x_q = torch.round(x / self.inp_scale).clamp(0, 15)
             
        # Step 2: Conv1 with quantized weights
        x_q = F.conv2d(x_q, self.weight_conv1, bias=None, stride=self.stride1, padding=self.pad1)
        
        # Step 3: Rescale conv1 output from "conv1 output scale" to quant_conv1 input scale
        scale_conv1_out = self.inp_scale * self.scale_conv1
        x_q = torch.round(x_q * (scale_conv1_out / self.act_scale_conv1)).clamp(0, 15)
    
        # Step 4: FC layer (1x1 conv) with quantized weights
        x_q = F.conv2d(x_q, self.weight_fc, bias=None, stride=1, padding=0)
    
        # Step 5: Pool and flatten
        x_q = F.adaptive_avg_pool2d(x_q, (1, 1))
        x_q = x_q.view(x_q.size(0), -1)
    
        # Step 6: Rescale final output to output quantization scale and clamp
        scale_out = self.act_scale_conv1 * self.scale_fc
        x_q = torch.round(x_q * (scale_out / self.out_scale)).clamp(0, 15)
    
        return x_q
```

**Comment by ajay-vikram** [comment]
okay thanks. one more doubt

quant_input = QuantIdentity() 
x= quant_input(x)
scale1 = x.scale
scale2 = model.quant_input.act_quant.fused_activation_quant_proxy.tensor_quant.scaling_impl.value.item()

why are these 2 scales different?

**Comment by ajay-vikram** [comment]
> Usually, this is done before quantization is applied, thus is has no effect on quantization since the new fused weights will be quantized instead of the "original" one

so during my training if I have a BatchNorm just after my QuantConv2d and before my QuantReLU, would I not need to do any normalization during real int inference? Is it enough if I fold the BN weights into the Conv2d and make the BN parameters frozen?

**Comment by Giuseppe5** [comment]
Hey, would you be needing more help with this?

**Comment by Giuseppe5** [comment]
Closing this, feel free to re-open in case you still have issues or questions.
